File: src/packages/agent_core/builtin_tools/web_fetch/executor.py
from __future__ import annotations

import time
from typing import Any, Callable

import anyio

from packages.agent_core.executor import (
    ToolExecutionContext,
    ToolExecutionError,
    ToolExecutionResult,
)
from packages.agent_core.tools import ToolSpec as AgentToolSpec
from packages.llm_gateway import ToolSpec as LlmToolSpec

from .basic import BasicWebFetchHttpError, BasicWebFetchProvider
from .config import WebFetchConfig
from .provider import WebFetchProvider, WebFetchResult
from .url_policy import UrlPolicyDeniedError, ensure_url_allowed

_ERROR_CLASS_ARGS_INVALID = "tool.args_invalid"
_ERROR_CLASS_TIMEOUT = "tool.timeout"
_ERROR_CLASS_FETCH_FAILED = "tool.fetch_failed"
_ERROR_CLASS_URL_DENIED = "tool.url_denied"
_ERROR_CLASS_NOT_CONFIGURED = "tool.not_configured"

_DEFAULT_TIMEOUT_SECONDS = 15.0
_MAX_LENGTH_LIMIT = 200_000

WEB_FETCH_AGENT_TOOL_SPEC = AgentToolSpec(
    name="web_fetch",
    version="1",
    description="抓取网页内容并提取正文",
    risk_level="medium",
    side_effects=False,
)

WEB_FETCH_LLM_TOOL_SPEC = LlmToolSpec(
    name="web_fetch",
    description="抓取网页内容，返回 title/content（纯文本）",
    json_schema={
        "type": "object",
        "properties": {
            "url": {"type": "string", "minLength": 1},
            "max_length": {"type": "integer", "minimum": 1, "maximum": _MAX_LENGTH_LIMIT},
        },
        "required": ["url", "max_length"],
        "additionalProperties": False,
    },
)


def _provider_from_env() -> WebFetchProvider | None:
    config = WebFetchConfig.from_env()
    if config is None:
        return None

    if config.provider_kind == "basic":
        return BasicWebFetchProvider()

    raise ValueError(f"web_fetch provider 未实现：{config.provider_kind}")


class WebFetchToolExecutor:
    def __init__(
        self,
        *,
        provider: WebFetchProvider | None = None,
        provider_factory: Callable[[], WebFetchProvider | None] | None = None,
        fallback_provider_factory: Callable[[], WebFetchProvider] | None = None,
        timeout_seconds: float = _DEFAULT_TIMEOUT_SECONDS,
    ) -> None:
        self._provider = provider
        self._provider_factory = provider_factory or _provider_from_env
        self._fallback_provider_factory = fallback_provider_factory or BasicWebFetchProvider
        self._timeout_seconds = float(timeout_seconds)

    async def execute(
        self,
        *,
        tool_name: str,
        args: dict[str, Any],
        context: ToolExecutionContext,
        tool_call_id: str | None = None,
    ) -> ToolExecutionResult:
        _ = (tool_name, context, tool_call_id)
        started = time.monotonic()
        url, max_length, error = _parse_web_fetch_args(args)
        if error is not None:
            return ToolExecutionResult(error=error, duration_ms=_duration_ms(started))

        provider = self._provider
        if provider is None:
            try:
                provider = self._provider_factory()
            except ValueError as exc:
                return ToolExecutionResult(
                    error=ToolExecutionError(
                        error_class=_ERROR_CLASS_NOT_CONFIGURED,
                        message="web_fetch 配置无效",
                        details={"reason": str(exc)},
                    ),
                    duration_ms=_duration_ms(started),
                )
            if provider is None:
                provider = self._fallback_provider_factory()

        try:
            ensure_url_allowed(url)
        except UrlPolicyDeniedError as exc:
            details: dict[str, object] = {"reason": exc.reason}
            if exc.details:
                details.update(dict(exc.details))
            return ToolExecutionResult(
                error=ToolExecutionError(
                    error_class=_ERROR_CLASS_URL_DENIED,
                    message="web_fetch URL 被安全策略拒绝",
                    details=details,
                ),
                duration_ms=_duration_ms(started),
            )

        try:
            with anyio.fail_after(self._timeout_seconds):
                result = await provider.fetch(url=url, max_length=max_length)
        except TimeoutError:
            return ToolExecutionResult(
                error=ToolExecutionError(
                    error_class=_ERROR_CLASS_TIMEOUT,
                    message="web_fetch 超时",
                    details={"timeout_seconds": self._timeout_seconds},
                ),
                duration_ms=_duration_ms(started),
            )
        except BasicWebFetchHttpError as exc:
            return ToolExecutionResult(
                error=ToolExecutionError(
                    error_class=_ERROR_CLASS_FETCH_FAILED,
                    message="web_fetch 请求失败",
                    details={"status_code": exc.status_code},
                ),
                duration_ms=_duration_ms(started),
            )
        except Exception as exc:
            return ToolExecutionResult(
                error=ToolExecutionError(
                    error_class=_ERROR_CLASS_FETCH_FAILED,
                    message="web_fetch 执行失败",
                    details={"exception_type": type(exc).__name__},
                ),
                duration_ms=_duration_ms(started),
            )

        if not isinstance(result, WebFetchResult):
            return ToolExecutionResult(
                error=ToolExecutionError(
                    error_class=_ERROR_CLASS_FETCH_FAILED,
                    message="web_fetch 返回结果类型不正确",
                ),
                duration_ms=_duration_ms(started),
            )

        return ToolExecutionResult(result_json=result.to_json(), duration_ms=_duration_ms(started))


def _parse_web_fetch_args(
    args: dict[str, Any],
) -> tuple[str | None, int | None, ToolExecutionError | None]:
    unknown = [key for key in args.keys() if key not in {"url", "max_length"}]
    if unknown:
        return None, None, ToolExecutionError(
            error_class=_ERROR_CLASS_ARGS_INVALID,
            message="工具参数不支持额外字段",
            details={"unknown_fields": sorted(unknown)},
        )

    url = args.get("url")
    if not isinstance(url, str) or not url.strip():
        return None, None, ToolExecutionError(
            error_class=_ERROR_CLASS_ARGS_INVALID,
            message="参数 url 必须为非空字符串",
            details={"field": "url"},
        )

    max_length = args.get("max_length")
    if not isinstance(max_length, int):
        return None, None, ToolExecutionError(
            error_class=_ERROR_CLASS_ARGS_INVALID,
            message="参数 max_length 必须为整数",
            details={"field": "max_length"},
        )
    if max_length <= 0 or max_length > _MAX_LENGTH_LIMIT:
        return None, None, ToolExecutionError(
            error_class=_ERROR_CLASS_ARGS_INVALID,
            message=f"参数 max_length 必须在 1..{_MAX_LENGTH_LIMIT} 之间",
            details={"field": "max_length", "max": _MAX_LENGTH_LIMIT},
        )

    return url.strip(), int(max_length), None


def _duration_ms(started: float) -> int:
    elapsed = time.monotonic() - started
    millis = int(elapsed * 1000)
    return millis if millis >= 0 else 0


__all__ = [
    "WEB_FETCH_AGENT_TOOL_SPEC",
    "WEB_FETCH_LLM_TOOL_SPEC",
    "WebFetchToolExecutor",
]

File: src/services/worker/internal/tools/builtin/web_fetch/executor.go
package webfetch

import (
	"context"
	"errors"
	"fmt"
	"net"
	"sort"
	"strings"
	"time"

	sharedtoolmeta "arkloop/services/shared/toolmeta"
	"arkloop/services/worker/internal/llm"
	"arkloop/services/worker/internal/tools"
)

const (
	errorArgsInvalid   = "tool.args_invalid"
	errorTimeout       = "tool.timeout"
	errorFetchFailed   = "tool.fetch_failed"
	errorURLDenied     = "tool.url_denied"
	errorNotConfigured = "config.missing"

	defaultTimeout = basicFetchTimeout
	maxLengthLimit = sharedtoolmeta.WebFetchMaxLengthLimit

	fetchFailureDNSFailed              = "dns_failed"
	fetchFailureEmptyPage              = "empty_page"
	fetchFailureHTTPStatusError        = "http_status_error"
	fetchFailureNetworkError           = "network_error"
	fetchFailureNetworkTimeout         = "network_timeout"
	fetchFailureRequestCanceled        = "request_canceled"
	fetchFailureTLSFailed              = "tls_failed"
	fetchFailureUnknown                = "unknown"
	fetchFailureUnsupportedContentType = "unsupported_content_type"
)

var AgentSpec = tools.AgentToolSpec{
	Name:        "web_fetch",
	Version:     "1",
	Description: "fetch web page content and extract body text",
	RiskLevel:   tools.RiskLevelMedium,
	SideEffects: false,
}

var AgentSpecJina = tools.AgentToolSpec{
	Name:        "web_fetch.jina",
	LlmName:     "web_fetch",
	Version:     "1",
	Description: "fetch web page content and extract body text",
	RiskLevel:   tools.RiskLevelMedium,
	SideEffects: false,
}

var AgentSpecFirecrawl = tools.AgentToolSpec{
	Name:        "web_fetch.firecrawl",
	LlmName:     "web_fetch",
	Version:     "1",
	Description: "fetch web page content and extract body text",
	RiskLevel:   tools.RiskLevelMedium,
	SideEffects: false,
}

var AgentSpecBasic = tools.AgentToolSpec{
	Name:        "web_fetch.basic",
	LlmName:     "web_fetch",
	Version:     "1",
	Description: "fetch web page content and extract body text",
	RiskLevel:   tools.RiskLevelMedium,
	SideEffects: false,
}

var LlmSpec = llm.ToolSpec{
	Name:        "web_fetch",
	Description: stringPtr(sharedtoolmeta.Must("web_fetch").LLMDescription),
	JSONSchema: map[string]any{
		"type": "object",
		"properties": map[string]any{
			"url":        map[string]any{"type": "string", "description": "URL to fetch"},
			"max_length": map[string]any{"type": "integer", "minimum": 1, "maximum": maxLengthLimit, "description": "maximum number of characters to return"},
		},
		"required":             []string{"url", "max_length"},
		"additionalProperties": false,
	},
}

type ToolExecutor struct {
	provider Provider
	timeout  time.Duration
}

func NewToolExecutor(_ any) *ToolExecutor {
	return &ToolExecutor{timeout: defaultTimeout}
}

func NewBasicExecutor(_ any) *ToolExecutor {
	return &ToolExecutor{timeout: defaultTimeout}
}

func NewFirecrawlExecutor(_ any) *ToolExecutor {
	return &ToolExecutor{timeout: defaultTimeout}
}

func NewJinaExecutor(_ any) *ToolExecutor {
	return &ToolExecutor{timeout: defaultTimeout}
}

func NewToolExecutorWithProvider(provider Provider) *ToolExecutor {
	return &ToolExecutor{
		provider: provider,
		timeout:  defaultTimeout,
	}
}

func (e *ToolExecutor) IsNotConfigured() bool {
	return e == nil || e.provider == nil
}

func (e *ToolExecutor) Execute(
	ctx context.Context,
	toolName string,
	args map[string]any,
	execCtx tools.ExecutionContext,
	_ string,
) tools.ExecutionResult {
	_ = toolName
	started := time.Now()

	targetURL, maxLength, argErr := parseArgs(args)
	if argErr != nil {
		return tools.ExecutionResult{
			Error:      argErr,
			DurationMs: durationMs(started),
		}
	}

	provider := e.provider
	if provider == nil {
		return tools.ExecutionResult{
			Error: &tools.ExecutionError{
				ErrorClass: errorNotConfigured,
				Message:    "web_fetch backend not configured",
			},
			DurationMs: durationMs(started),
		}
	}

	if err := EnsureURLAllowed(targetURL); err != nil {
		denied, ok := err.(UrlPolicyDeniedError)
		details := map[string]any{"reason": "unknown"}
		if ok {
			details = map[string]any{"reason": denied.Reason}
			for key, value := range denied.Details {
				details[key] = value
			}
		}
		return tools.ExecutionResult{
			Error: &tools.ExecutionError{
				ErrorClass: errorURLDenied,
				Message:    "web_fetch URL denied by security policy",
				Details:    details,
			},
			DurationMs: durationMs(started),
		}
	}

	timeout := e.timeout
	if execCtx.TimeoutMs != nil && *execCtx.TimeoutMs > 0 {
		timeout = time.Duration(*execCtx.TimeoutMs) * time.Millisecond
	}
	fetchCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	result, err := provider.Fetch(fetchCtx, targetURL, maxLength)
	if err != nil {
		var denied UrlPolicyDeniedError
		if errors.As(err, &denied) {
			details := map[string]any{"reason": denied.Reason}
			for key, value := range denied.Details {
				details[key] = value
			}
			return tools.ExecutionResult{
				Error: &tools.ExecutionError{
					ErrorClass: errorURLDenied,
					Message:    "web_fetch URL denied by security policy",
					Details:    details,
				},
				DurationMs: durationMs(started),
			}
		}
		if errors.Is(err, context.DeadlineExceeded) {
			return tools.ExecutionResult{
				Error: &tools.ExecutionError{
					ErrorClass: errorTimeout,
					Message:    "web_fetch timed out",
					Details: map[string]any{
						"reason":          fetchFailureNetworkTimeout,
						"retryable":       true,
						"timeout_seconds": timeout.Seconds(),
					},
				},
				DurationMs: durationMs(started),
			}
		}
		if errors.Is(err, context.Canceled) {
			return tools.ExecutionResult{
				Error: &tools.ExecutionError{
					ErrorClass: errorFetchFailed,
					Message:    "web_fetch request canceled",
					Details: map[string]any{
						"reason":    fetchFailureRequestCanceled,
						"retryable": false,
					},
				},
				DurationMs: durationMs(started),
			}
		}
		var httpErr HttpError
		if errors.As(err, &httpErr) {
			return tools.ExecutionResult{
				Error: &tools.ExecutionError{
					ErrorClass: errorFetchFailed,
					Message:    "web_fetch request failed",
					Details: map[string]any{
						"reason":      fetchFailureHTTPStatusError,
						"retryable":   isRetryableHTTPStatus(httpErr.StatusCode),
						"status_code": httpErr.StatusCode,
					},
				},
				DurationMs: durationMs(started),
			}
		}
		var contentTypeErr UnsupportedContentTypeError
		if errors.As(err, &contentTypeErr) {
			return tools.ExecutionResult{
				Error: &tools.ExecutionError{
					ErrorClass: errorFetchFailed,
					Message:    "web_fetch content type is not readable text",
					Details: map[string]any{
						"reason":       fetchFailureUnsupportedContentType,
						"retryable":    false,
						"content_type": contentTypeErr.ContentType,
					},
				},
				DurationMs: durationMs(started),
			}
		}
		failure := classifyFetchFailure(err)
		return tools.ExecutionResult{
			Error: &tools.ExecutionError{
				ErrorClass: errorFetchFailed,
				Message:    "web_fetch execution failed",
				Details: map[string]any{
					"reason":    failure.reason,
					"retryable": failure.retryable,
					"error":     err.Error(),
				},
			},
			DurationMs: durationMs(started),
		}
	}

	return tools.ExecutionResult{
		ResultJSON: result.ToJSON(),
		DurationMs: durationMs(started),
	}
}

type fetchFailure struct {
	reason    string
	retryable bool
}

func classifyFetchFailure(err error) fetchFailure {
	if err == nil {
		return fetchFailure{reason: fetchFailureUnknown}
	}
	if errors.Is(err, context.Canceled) {
		return fetchFailure{reason: fetchFailureRequestCanceled}
	}
	if errors.Is(err, context.DeadlineExceeded) {
		return fetchFailure{reason: fetchFailureNetworkTimeout, retryable: true}
	}

	var dnsErr *net.DNSError
	if errors.As(err, &dnsErr) {
		return fetchFailure{reason: fetchFailureDNSFailed, retryable: dnsErr.IsTimeout || dnsErr.IsTemporary}
	}

	var netErr net.Error
	if errors.As(err, &netErr) {
		if netErr.Timeout() {
			return fetchFailure{reason: fetchFailureNetworkTimeout, retryable: true}
		}
	}

	message := strings.ToLower(err.Error())
	switch {
	case strings.Contains(message, "outbound dns resolve failed"),
		strings.Contains(message, "no such host"),
		strings.Contains(message, "server misbehaving"):
		return fetchFailure{reason: fetchFailureDNSFailed, retryable: true}
	case strings.Contains(message, "tls handshake timeout"),
		strings.Contains(message, "tls:"),
		strings.Contains(message, "certificate"),
		strings.Contains(message, "x509:"):
		return fetchFailure{reason: fetchFailureTLSFailed, retryable: strings.Contains(message, "timeout")}
	case strings.Contains(message, "timeout"),
		strings.Contains(message, "deadline exceeded"),
		strings.Contains(message, "i/o timeout"):
		return fetchFailure{reason: fetchFailureNetworkTimeout, retryable: true}
	case strings.Contains(message, "connection reset"),
		strings.Contains(message, "connection refused"),
		strings.Contains(message, "unexpected eof"),
		strings.Contains(message, "eof"),
		strings.Contains(message, "temporary failure"),
		strings.Contains(message, "try again"):
		return fetchFailure{reason: fetchFailureNetworkError, retryable: true}
	case strings.Contains(message, "response body is empty"),
		strings.Contains(message, "html-empty"):
		return fetchFailure{reason: fetchFailureEmptyPage}
	default:
		return fetchFailure{reason: fetchFailureUnknown}
	}
}

func isRetryableHTTPStatus(statusCode int) bool {
	return statusCode == 408 || statusCode == 429 || statusCode == 502 || statusCode == 503 || statusCode == 504
}

func parseArgs(args map[string]any) (string, int, *tools.ExecutionError) {
	unknown := []string{}
	for key := range args {
		if key != "url" && key != "max_length" {
			unknown = append(unknown, key)
		}
	}
	if len(unknown) > 0 {
		sort.Strings(unknown)
		return "", 0, &tools.ExecutionError{
			ErrorClass: errorArgsInvalid,
			Message:    "tool arguments do not allow extra fields",
			Details:    map[string]any{"unknown_fields": unknown},
		}
	}

	rawURL, ok := args["url"].(string)
	if !ok || strings.TrimSpace(rawURL) == "" {
		return "", 0, &tools.ExecutionError{
			ErrorClass: errorArgsInvalid,
			Message:    "parameter url must be a non-empty string",
			Details:    map[string]any{"field": "url"},
		}
	}
	targetURL := normalizeTargetURL(rawURL)

	rawMax, ok := args["max_length"]
	maxLength, okInt := rawMax.(int)
	if !ok || !okInt {
		if floatVal, ok := rawMax.(float64); ok {
			maxLength = int(floatVal)
			okInt = floatVal == float64(maxLength)
		}
	}
	if !okInt {
		return "", 0, &tools.ExecutionError{
			ErrorClass: errorArgsInvalid,
			Message:    "parameter max_length must be an integer",
			Details:    map[string]any{"field": "max_length"},
		}
	}
	if maxLength <= 0 || maxLength > maxLengthLimit {
		return "", 0, &tools.ExecutionError{
			ErrorClass: errorArgsInvalid,
			Message:    fmt.Sprintf("parameter max_length must be in range 1..%d", maxLengthLimit),
			Details:    map[string]any{"field": "max_length", "max": maxLengthLimit},
		}
	}
	return targetURL, maxLength, nil
}

func normalizeTargetURL(raw string) string {
	cleaned := strings.TrimSpace(raw)
	if cleaned == "" {
		return ""
	}

	cleaned = fixDuplicatedScheme(cleaned)
	cleaned = unwrapJinaWrapper(cleaned)
	cleaned = fixDuplicatedScheme(cleaned)
	return strings.TrimSpace(cleaned)
}

func fixDuplicatedScheme(raw string) string {
	if strings.HasPrefix(raw, "httpshttps://") {
		return "https://" + strings.TrimPrefix(raw, "httpshttps://")
	}
	if strings.HasPrefix(raw, "httphttp://") {
		return "http://" + strings.TrimPrefix(raw, "httphttp://")
	}
	return raw
}

func unwrapJinaWrapper(raw string) string {
	trimmed := strings.TrimSpace(raw)
	for {
		stripped := false
		for _, prefix := range []string{"https://r.jina.ai/", "http://r.jina.ai/"} {
			if strings.HasPrefix(trimmed, prefix) {
				trimmed = strings.TrimSpace(strings.TrimPrefix(trimmed, prefix))
				stripped = true
			}
		}
		if !stripped {
			break
		}
	}
	return trimmed
}

func stringPtr(value string) *string {
	cleaned := strings.TrimSpace(value)
	if cleaned == "" {
		return nil
	}
	return &cleaned
}

func durationMs(started time.Time) int {
	elapsed := time.Since(started)
	millis := int(elapsed / time.Millisecond)
	if millis < 0 {
		return 0
	}
	return millis
}

// jinaWithBasicFallback tries Jina (anonymous free tier) first and silently
// falls back to the basic HTTP provider on any non-policy, non-timeout error.
type jinaWithBasicFallback struct {
	jina  Provider
	basic Provider
}

func (p *jinaWithBasicFallback) Fetch(ctx context.Context, url string, maxLength int) (Result, error) {
	result, err := p.jina.Fetch(ctx, url, maxLength)
	if err == nil {
		return result, nil
	}
	// Do not fall back for URL policy violations or context cancellation —
	// those errors would be the same with basic.
	var policyErr UrlPolicyDeniedError
	if errors.As(err, &policyErr) {
		return Result{}, err
	}
	if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
		return Result{}, err
	}
	return p.basic.Fetch(ctx, url, maxLength)
}
